### ISSUE TITLE
Uptake BucketUpdater for upgrade, stability

### ISSUE THREAD

**Issue by adamcfraser** [couchbase/sync_gateway#1011]
go-couchbase has added support for listening to the bucketsStreaming endpoint to automatically respond to updated cluster configurations published by ns_server.  Among other things, this should allow Sync Gateway clients to gracefully handle the upgrade via rebalance scenario.

Details from Manik:

```
I believe that go-couchbase already might have something that solves your
problem.  Please have a look at the example here :
https://github.com/couchbase/go-couchbase/blob/master/examples/streaming/ex
ample.go
Once you connect to a serverURL and create a reference to a bucket, you
need to call the method RunBucketUpdater with a  callback function.
RunBucketUpdater internally connects to one of the nodes in the cluster
and listens for vbucket updates. Even if there is a network error it will
update the bucket information from one of the servers that are still
available.  From the clients perspective the bucket reference returned in
the GetBucket call remains valid and it can continue to perform operations
on the bucket.
I ran the example above with the following params :
./streaming -bucketName="beer-sample" -serverURL="http://localhost:9000"
And subsequently brought down the server on localhost:9000, I saw that the
vbucket configuration was updated and after I clicked on rebalance the set
operations continued normally.
```

**Comment by adamcfraser** [comment]
Needs a perf test to validate the change.

**Comment by adamcfraser** [comment]
Add warning in the Sync Gateway log when we get the bucketUpdater callback, to recommend a restart (until 1.2.0 handles offline/online automatically).

**Comment by adamcfraser** [comment]
Standard performance run was clean:
http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/960/

**Comment by adamcfraser** [comment]
Merged to master by #1028, 1.1.1 by #1029.
